Hum ... this is failing with VS2015 (which is not in our CI system):

```
C:\PROGRA~2\MICROS~1.0\VC\bin\amd64\cl.exe  /nologo /TP -DNOMINMAX -DWIN32_LEAN_AND_MEAN -Dexiv2lib_STATIC -I. -Isrc -I"C:\Users\Luis Diaz\.conan\data\zlib\1.2.11\conan\stable\package\63da998e3642b50bee33f4449826b2d623661505\include" -I..\include\exiv2 /DWIN32 /D_WINDOWS /W3 /GR /EHsc  /MP /MDd /Zi /Ob0 /Od /RTC1 /showIncludes /Fosrc\CMakeFiles\exiv2lib_int.dir\crwimage_int.cpp.obj /Fdsrc\CMakeFiles\exiv2lib_int.dir\ /FS -c ..\src\crwimage_int.cpp
..\src\crwimage_int.cpp(155): error C2373: 'signature_': redefinition; different type modifiers
c:\dev\exiv2\src\crwimage_int.hpp(518): note: see declaration of 'signature_'
```

I am afraid we could be using something from C++14 or C++17 for the Visual Studio case without noticing ... take a look to this post:
https://devblogs.microsoft.com/cppblog/standards-version-switches-in-the-compiler/

I wonder if we should add a VS2015 image in Appveyor to control these cases.

--- a/src/crwimage_int.cpp
+++ b/src/crwimage_int.cpp
@@ -152,7 +152,7 @@ namespace Exiv2 {
         { 0xffff, 0xffff }
     };
 
-    const char CiffHeader::signature_[] = "HEAPCCDR";
+    constexpr char CiffHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
     CiffHeader::~CiffHeader()
     {

--- a/src/crwimage_int.hpp
+++ b/src/crwimage_int.hpp
@@ -514,7 +514,8 @@ namespace Exiv2 {
 
     private:
         // DATA
-        static const char signature_[];   //!< Canon CRW signature "HEAPCCDR"
+        static constexpr char signature_[]{
+            'H', 'E', 'A', 'P', 'C', 'C', 'D', 'R', '\0' };   //!< Canon CRW signature "HEAPCCDR"
 
         CiffDirectory*    pRootDir_;      //!< Pointer to the root directory
         ByteOrder         byteOrder_;     //!< Applicable byte order

--- a/src/makernote_int.cpp
+++ b/src/makernote_int.cpp
@@ -227,11 +227,9 @@ namespace Exiv2 {
         return 0;
     }
 
-    const byte OlympusMnHeader::signature_[] = {
-        'O', 'L', 'Y', 'M', 'P', 0x00, 0x01, 0x00
-    };
+    constexpr byte OlympusMnHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
-    uint32_t OlympusMnHeader::sizeOfSignature()
+    constexpr uint32_t OlympusMnHeader::sizeOfSignature()
     {
         return sizeof(signature_);
     }
@@ -276,11 +274,9 @@ namespace Exiv2 {
         return sizeOfSignature();
     } // OlympusMnHeader::write
 
-    const byte Olympus2MnHeader::signature_[] = {
-        'O', 'L', 'Y', 'M', 'P', 'U', 'S', 0x00, 'I', 'I', 0x03, 0x00
-    };
+    constexpr byte Olympus2MnHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
-    uint32_t Olympus2MnHeader::sizeOfSignature()
+    constexpr uint32_t Olympus2MnHeader::sizeOfSignature()
     {
         return sizeof(signature_);
     }
@@ -330,12 +326,10 @@ namespace Exiv2 {
         return sizeOfSignature();
     } // Olympus2MnHeader::write
 
-    const byte FujiMnHeader::signature_[] = {
-        'F', 'U', 'J', 'I', 'F', 'I', 'L', 'M', 0x0c, 0x00, 0x00, 0x00
-    };
-    const ByteOrder FujiMnHeader::byteOrder_ = littleEndian;
+    constexpr byte FujiMnHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
+    constexpr ByteOrder FujiMnHeader::byteOrder_; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
-    uint32_t FujiMnHeader::sizeOfSignature()
+    constexpr uint32_t FujiMnHeader::sizeOfSignature()
     {
         return sizeof(signature_);
     }
@@ -393,11 +387,9 @@ namespace Exiv2 {
         return sizeOfSignature();
     } // FujiMnHeader::write
 
-    const byte Nikon2MnHeader::signature_[] = {
-        'N', 'i', 'k', 'o', 'n', '\0', 0x01, 0x00
-    };
+    constexpr byte Nikon2MnHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
-    uint32_t Nikon2MnHeader::sizeOfSignature()
+    constexpr uint32_t Nikon2MnHeader::sizeOfSignature()
     {
         return sizeof(signature_);
     }
@@ -440,12 +432,9 @@ namespace Exiv2 {
         return sizeOfSignature();
     } // Nikon2MnHeader::write
 
-    const byte Nikon3MnHeader::signature_[] = {
-        'N', 'i', 'k', 'o', 'n', '\0', 0x02, 0x10, 0x00, 0x00,
-        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
-    };
+    constexpr byte Nikon3MnHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
-    uint32_t Nikon3MnHeader::sizeOfSignature()
+    constexpr uint32_t Nikon3MnHeader::sizeOfSignature()
     {
         return sizeof(signature_);
     }
@@ -516,11 +505,9 @@ namespace Exiv2 {
         byteOrder_ = byteOrder;
     }
 
-    const byte PanasonicMnHeader::signature_[] = {
-        'P', 'a', 'n', 'a', 's', 'o', 'n', 'i', 'c', 0x00, 0x00, 0x00
-    };
+    constexpr byte PanasonicMnHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
-    uint32_t PanasonicMnHeader::sizeOfSignature()
+    constexpr uint32_t PanasonicMnHeader::sizeOfSignature()
     {
         return sizeof(signature_);
     }
@@ -563,11 +550,8 @@ namespace Exiv2 {
         return sizeOfSignature();
     } // PanasonicMnHeader::write
 
-    const byte PentaxDngMnHeader::signature_[] = {
-        'P', 'E', 'N', 'T', 'A', 'X', ' ', 0x00, 'M', 'M'
-    };
-
-    uint32_t PentaxDngMnHeader::sizeOfSignature()
+    constexpr byte PentaxDngMnHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
+    constexpr uint32_t PentaxDngMnHeader::sizeOfSignature()
     {
         return sizeof(signature_);
     }
@@ -617,11 +601,9 @@ namespace Exiv2 {
         return sizeOfSignature();
     } // PentaxDngMnHeader::write
 
-    const byte PentaxMnHeader::signature_[] = {
-        'A', 'O', 'C', 0x00, 'M', 'M'
-    };
+    constexpr byte PentaxMnHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
-    uint32_t PentaxMnHeader::sizeOfSignature()
+    constexpr uint32_t PentaxMnHeader::sizeOfSignature()
     {
         return sizeof(signature_);
     }
@@ -694,16 +676,12 @@ namespace Exiv2 {
         return 0;
     } // SamsungMnHeader::write
 
-    const byte SigmaMnHeader::signature1_[] = {
-        'S', 'I', 'G', 'M', 'A', '\0', '\0', '\0', 0x01, 0x00
-    };
-    const byte SigmaMnHeader::signature2_[] = {
-        'F', 'O', 'V', 'E', 'O', 'N', '\0', '\0', 0x01, 0x00
-    };
+    constexpr byte SigmaMnHeader::signature1_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
+    constexpr byte SigmaMnHeader::signature2_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
-    uint32_t SigmaMnHeader::sizeOfSignature()
+    constexpr uint32_t SigmaMnHeader::sizeOfSignature()
     {
-        assert(sizeof(signature1_) == sizeof(signature2_));
+        static_assert(sizeof(signature1_) == sizeof(signature2_), "Signatures must be of same size");
         return sizeof(signature1_);
     }
 
@@ -746,11 +724,9 @@ namespace Exiv2 {
         return sizeOfSignature();
     } // SigmaMnHeader::write
 
-    const byte SonyMnHeader::signature_[] = {
-        'S', 'O', 'N', 'Y', ' ', 'D', 'S', 'C', ' ', '\0', '\0', '\0'
-    };
+    constexpr byte SonyMnHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
-    uint32_t SonyMnHeader::sizeOfSignature()
+    constexpr uint32_t SonyMnHeader::sizeOfSignature()
     {
         return sizeof(signature_);
     }
@@ -793,12 +769,10 @@ namespace Exiv2 {
         return sizeOfSignature();
     } // SonyMnHeader::write
 
-    const byte Casio2MnHeader::signature_[] = {
-        'Q', 'V', 'C', '\0', '\0', '\0'
-    };
-    const ByteOrder Casio2MnHeader::byteOrder_ = bigEndian;
+    constexpr byte Casio2MnHeader::signature_[]; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
+    constexpr ByteOrder Casio2MnHeader::byteOrder_; /// \todo Required in C++11, but deprecated in C++17. Remove this line when we move to C++17.
 
-    uint32_t Casio2MnHeader::sizeOfSignature()
+    constexpr uint32_t Casio2MnHeader::sizeOfSignature()
     {
         return sizeof(signature_);
     }

--- a/src/makernote_int.hpp
+++ b/src/makernote_int.hpp
@@ -200,11 +200,12 @@ namespace Exiv2 {
         uint32_t ifdOffset() const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf header_;                //!< Data buffer for the makernote header
-        static const byte signature_[]; //!< Olympus makernote header signature
+        static constexpr byte signature_[]{
+            'O', 'L', 'Y', 'M', 'P', 0x00, 0x01, 0x00 }; //!< Olympus makernote header signature
 
     }; // class OlympusMnHeader
 
@@ -230,11 +231,12 @@ namespace Exiv2 {
         uint32_t baseOffset(uint32_t mnOffset) const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf header_;                //!< Data buffer for the makernote header
-        static const byte signature_[]; //!< Olympus makernote header signature
+        static constexpr byte signature_[]{
+            'O', 'L', 'Y', 'M', 'P', 'U', 'S', 0x00, 'I', 'I', 0x03, 0x00 }; //!< Olympus makernote header signature
 
     }; // class Olympus2MnHeader
 
@@ -262,12 +264,13 @@ namespace Exiv2 {
         uint32_t baseOffset(uint32_t mnOffset) const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf header_;                //!< Data buffer for the makernote header
-        static const byte signature_[]; //!< Fujifilm makernote header signature
-        static const ByteOrder byteOrder_; //!< Byteorder for makernote (always II)
+        static constexpr byte signature_[]{
+            'F', 'U', 'J', 'I', 'F', 'I', 'L', 'M', 0x0c, 0x00, 0x00, 0x00 }; //!< Fujifilm makernote header signature
+        static constexpr ByteOrder byteOrder_{ littleEndian }; //!< Byteorder for makernote (always II)
         uint32_t start_;                //!< Start of the mn IFD rel. to mn start
 
     }; // class FujiMnHeader
@@ -293,12 +296,13 @@ namespace Exiv2 {
         uint32_t ifdOffset() const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf buf_;                   //!< Raw header data
         uint32_t start_;                //!< Start of the mn IFD rel. to mn start
-        static const byte signature_[]; //!< Nikon 2 makernote header signature
+        static constexpr byte signature_[]{
+            'N', 'i', 'k', 'o', 'n', '\0', 0x01, 0x00 }; //!< Nikon 2 makernote header signature
 
     }; // class Nikon2MnHeader
 
@@ -326,13 +330,15 @@ namespace Exiv2 {
         uint32_t baseOffset(uint32_t mnOffset) const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf buf_;                   //!< Raw header data
         ByteOrder byteOrder_;           //!< Byteorder for makernote
         uint32_t start_;                //!< Start of the mn IFD rel. to mn start
-        static const byte signature_[]; //!< Nikon 3 makernote header signature
+        static constexpr byte signature_[]{
+           'N', 'i', 'k', 'o', 'n', '\0', 0x02, 0x10, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }; //!< Nikon 3 makernote header signature
 
     }; // class Nikon3MnHeader
 
@@ -357,12 +363,13 @@ namespace Exiv2 {
         uint32_t ifdOffset() const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf buf_;                   //!< Raw header data
         uint32_t start_;                //!< Start of the mn IFD rel. to mn start
-        static const byte signature_[]; //!< Panasonic makernote header signature
+        static constexpr byte signature_[]{
+            'P', 'a', 'n', 'a', 's', 'o', 'n', 'i', 'c', 0x00, 0x00, 0x00 }; //!< Panasonic makernote header signature
 
     }; // class PanasonicMnHeader
 
@@ -388,11 +395,12 @@ namespace Exiv2 {
         uint32_t baseOffset(uint32_t mnOffset) const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf header_;                //!< Data buffer for the makernote header
-        static const byte signature_[]; //!< Pentax DNG makernote header signature
+        static constexpr byte signature_[]{
+            'P', 'E', 'N', 'T', 'A', 'X', ' ', 0x00, 'M', 'M' }; //!< Pentax DNG makernote header signature
 
     }; // class PentaxDngMnHeader
 
@@ -417,11 +425,12 @@ namespace Exiv2 {
         uint32_t ifdOffset() const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf header_;                //!< Data buffer for the makernote header
-        static const byte signature_[]; //!< Pentax makernote header signature
+        static constexpr byte signature_[]{
+            'A', 'O', 'C', 0x00, 'M', 'M' }; //!< Pentax makernote header signature
 
     }; // class PentaxMnHeader
 
@@ -467,13 +476,15 @@ namespace Exiv2 {
         uint32_t ifdOffset() const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf buf_;                    //!< Raw header data
         uint32_t start_;                 //!< Start of the mn IFD rel. to mn start
-        static const byte signature1_[]; //!< Sigma makernote header signature 1
-        static const byte signature2_[]; //!< Sigma makernote header signature 2
+        static constexpr byte signature1_[]{
+            'S', 'I', 'G', 'M', 'A', '\0', '\0', '\0', 0x01, 0x00 };  //!< Sigma makernote header signature 1
+        static constexpr byte signature2_[]{
+            'F', 'O', 'V', 'E', 'O', 'N', '\0', '\0', 0x01, 0x00 }; //!< Sigma makernote header signature 2
 
     }; // class SigmaMnHeader
 
@@ -498,12 +509,13 @@ namespace Exiv2 {
         uint32_t ifdOffset() const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf buf_;                   //!< Raw header data
         uint32_t start_;                //!< Start of the mn IFD rel. to mn start
-        static const byte signature_[]; //!< Sony makernote header signature
+        static constexpr byte signature_[]{
+            'S', 'O', 'N', 'Y', ' ', 'D', 'S', 'C', ' ', '\0', '\0', '\0' }; //!< Sony makernote header signature
 
     }; // class SonyMnHeader
 
@@ -529,13 +541,14 @@ namespace Exiv2 {
         ByteOrder byteOrder() const override;
         //@}
         //! Return the size of the makernote header signature
-        static uint32_t sizeOfSignature();
+        static constexpr uint32_t sizeOfSignature();
 
     private:
         DataBuf buf_;                   //!< Raw header data
         uint32_t start_;                //!< Start of the mn IFD rel. to mn start
-        static const byte signature_[]; //!< Casio makernote header signature
-        static const ByteOrder byteOrder_; //!< Byteorder for makernote (always big endian)
+        static constexpr byte signature_[]{
+            'Q', 'V', 'C', '\0', '\0', '\0' }; //!< Casio makernote header signature
+        static constexpr ByteOrder byteOrder_{ bigEndian }; //!< Byteorder for makernote (always big endian)
 
     }; // class Casio2MnHeader
 

--- a/src/types.cpp
+++ b/src/types.cpp
@@ -349,7 +349,7 @@ namespace Exiv2 {
         // This algorithm assumes that the internal representation of the float
         // type is the 4-byte IEEE 754 binary32 format, which is common but not
         // required by the C++ standard.
-        assert(sizeof(float) == 4);
+        static_assert(sizeof(float) == 4, "float type requires 4-byte IEEE 754 binary32 format");
         union {
             uint32_t ul_;
             float    f_;
@@ -363,7 +363,7 @@ namespace Exiv2 {
         // This algorithm assumes that the internal representation of the double
         // type is the 8-byte IEEE 754 binary64 format, which is common but not
         // required by the C++ standard.
-        assert(sizeof(double) == 8);
+        static_assert(sizeof(double) == 8, "double type requires 8-byte IEEE 754 binary64 format");
         union {
             uint64_t ull_;
             double   d_;
@@ -471,7 +471,7 @@ namespace Exiv2 {
         // This algorithm assumes that the internal representation of the float
         // type is the 4-byte IEEE 754 binary32 format, which is common but not
         // required by the C++ standard.
-        assert(sizeof(float) == 4);
+        static_assert(sizeof(float) == 4, "float type requires 4-byte IEEE 754 binary32 format");
         union {
             uint32_t ul_;
             float    f_;
@@ -485,7 +485,7 @@ namespace Exiv2 {
         // This algorithm assumes that the internal representation of the double
         // type is the 8-byte IEEE 754 binary64 format, which is common but not
         // required by the C++ standard.
-        assert(sizeof(double) == 8);
+        static_assert(sizeof(double) == 8, "double type requires 8-byte IEEE 754 binary64 format");
         union {
             uint64_t ull_;
             double   d_;

